src/library/cmd_reader.h

```diff
@@ -157,19 +166,305 @@ namespace xlang::cmd
-                    throw_invalid("Path '", path, "' is not a file or directory");
+                    std::regex rx(R"(((\d+)\.(\d+)\.(\d+)\.(\d+))\+?)");
+                    std::smatch match;
+
+                    if (std::regex_match(path, match, rx))
+                    {
+                        sdk_version = match[1].str();
+                    }
                 }
+
+                if (!sdk_version.empty())
+                {
+                    auto sdk_path = get_sdk_path();
+                    auto xml_path = sdk_path;
+                    xml_path /= L"Platforms\\UAP";
+                    xml_path /= sdk_version;
+                    xml_path /= L"Platform.xml";
+
+                    add_files_from_xml(files, sdk_version, xml_path, sdk_path);
+
+                    if (path.back() != '+')
+                    {
+                        continue;
+                    }
+
+                    for (auto&& item : std::experimental::filesystem::directory_iterator(sdk_path / L"Extension SDKs"))
+                    {
+                        xml_path = item.path() / sdk_version;
+                        xml_path /= L"SDKManifest.xml";
+
+                        add_files_from_xml(files, sdk_version, xml_path, sdk_path);
+                    }
+
+                    continue;
+                }
+#endif
+                throw_invalid("Path '", path, "' is not a file or directory");
             }
 
             return files;
         }
 
+        auto files(std::string_view const& name) const
+        {
+            return files(name, [](auto&&) {return true; });
+        }
+
     private:
 
+#if XLANG_PLATFORM_WINDOWS
+
+        struct registry_key
+        {
+            HKEY handle{};
+
+            registry_key(registry_key const&) = delete;
+            registry_key& operator=(registry_key const&) = delete;
+
+            ~registry_key() noexcept
+            {
+                if (handle)
+                {
+                    RegCloseKey(handle);
+                }
+            }
+        };
+
+        template <typename T>
+        struct com_ptr
+        {
+            T* ptr{};
+
+            com_ptr(com_ptr const&) = delete;
+            com_ptr& operator=(com_ptr const&) = delete;
+
+            com_ptr() noexcept = default;
+
+            ~com_ptr() noexcept
+            {
+                if (ptr)
+                {
+                    ptr->Release();
+                }
+            }
+
+            auto operator->() const noexcept
+            {
+                return ptr;
+            }
+        };
+
+        static void check_xml(HRESULT result)
+        {
+            if (result < 0)
+            {
+                throw_invalid("Could not read the Windows SDK's Platform.xml");
+            }
+        }
+
+        static void add_files_from_xml(
+            std::set<std::string>& files,
+            std::string const& sdk_version,
+            std::experimental::filesystem::path const& xml_path,
```

Comment:
should we be using std::filesystem::path now?

Reply:
Yes, I'd love to. I thought Harry mentioned some reason why that didn't work on the Ubuntu build at present. I can check whether this has been fixed.

Reply:
Clang 6.0's libc++ doesn't include filesystem in std namespace. https://github.com/llvm-mirror/libcxx/tree/release_60/include

Reply:
Clang 7 has been out for almost 6 months. 😉